app/src/main/java/de/tum/in/tumcampusapp/widgets/MVVWidget.java

```diff
@@ -0,0 +1,146 @@
+import android.app.PendingIntent;
+import android.appwidget.AppWidgetManager;
+import android.appwidget.AppWidgetProvider;
+import android.content.ComponentName;
+import android.content.Context;
+import android.content.Intent;
+import android.database.Cursor;
+import android.net.Uri;
+import android.support.annotation.NonNull;
+import android.widget.RemoteViews;
+
+import de.tum.in.tumcampusapp.R;
+import de.tum.in.tumcampusapp.managers.TransportManager;
+
+/**
+ * Implementation of App Widget functionality.
+ * App Widget Configuration implemented in {@link MVVWidgetConfigureActivity MVVWidgetConfigureActivity}
+ */
+public class MVVWidget extends AppWidgetProvider {
+
+    private static final String BROADCAST_ALARM_NAME = "de.tum.in.newtumcampus.intent.action.BROADCAST_MVVWIDGET_ALARM";
+    public static final String EXTRA_STATION_ID = "de.tum.in.newtumcampus.intent.action.MVV_WIDGET_EXTRA_STATION_ID";
+    private static boolean alarmIsSet = false;
+
+    @Override
+    public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds) {
+        for (int appWidgetId : appWidgetIds) {
+            updateAppWidget(context, appWidgetManager, appWidgetId);
+        }
+        setAlarm(context);
+        super.onUpdate(context, appWidgetManager, appWidgetIds);
+    }
+
+    @Override
+    public void onDeleted(Context context, int[] appWidgetIds) {
+        // When the user deletes the widget, delete the associated setting from the database.
+        TransportManager transportManager = new TransportManager(context);
+        for (int appWidgetId : appWidgetIds) {
+            transportManager.deleteWidget(appWidgetId);
+        }
+        super.onDeleted(context, appWidgetIds);
+    }
+
+    @Override
+    public void onEnabled(Context context) {
+        // Enter relevant functionality for when the first widget is created
+        setAlarm(context);
+    }
+
+    @Override
+    public void onDisabled(Context context) {
+        // Cancel alarm as the last widget has been removed
+        Intent intent = new Intent(context, MVVWidget.class);
+        PendingIntent sender = PendingIntent.getBroadcast(context, 0, intent, 0);
+        AlarmManager alarmManager = (AlarmManager) context.getSystemService(Context.ALARM_SERVICE);
+        alarmManager.cancel(sender);
+        alarmIsSet = false;
+        super.onDisabled(context);
+    }
+
+    /**
+     * If no alarm is running yet a new repeating alarm is started
+     */
+    private static void setAlarm(Context context) {
+        if (alarmIsSet) return;
+        AlarmManager am = (AlarmManager) context.getSystemService(Context.ALARM_SERVICE);
+        Intent intent = new Intent(context, MVVWidget.class);
+        intent.setAction(BROADCAST_ALARM_NAME);
+        PendingIntent pi = PendingIntent.getBroadcast(context, 0, intent, 0);
+        //After after 30 or 60 (android may delay) seconds
+        am.setRepeating(AlarmManager.RTC, System.currentTimeMillis(), 30000, pi);
```

Comment:
So this is a litte bit tricky
1) the usual onUpdate method of a widget is called every hour, at most, so this does not work for our usecase
2) using an alarm (as i do) has a min interval with about one minute (but can be delayed) and is not called when the device is not unlocked (which we want)
3) in every alarm step i use a time (which is exact) to update the widget after 30s and 60s to prevent late alarm steps -> at least every minute is one update (only with unlocked device)
4) i do download data only every second minute (at most) to reduce data and battery usage, between that the information is cached and the departure times are calculated
So overall we have downloadInterval, alarmInterval and 2x timerInterval to be configured... what would you suppose?